File: internal/gitstatus/gitstatus.go
package gitstatus

import (
	"os/exec"
	"strings"

	"skill-sync-manager/internal/models"
	"skill-sync-manager/internal/proc"
)

func runGit(args ...string) (string, error) {
	out, err := proc.RunCombined("proc:git", "", "git", args...)
	if err != nil {
		return out, err
	}
	return out, nil
}

// IsGitAvailable 判断本机是否能正常调用 `git --version`。
func IsGitAvailable() bool {
	_, err := exec.LookPath("git")
	return err == nil
}

// GetGitStatus 针对 root 运行一系列只读的 git 命令，返回稳定的结构体。
// 命令失败不会作为错误抛出 —— 前端看到的是对应字段被置为 false。
func GetGitStatus(root string) models.GitStatus {
	status := models.GitStatus{
		Remotes: []string{},
	}

	if root == "" {
		return status
	}

	status.GitAvailable = IsGitAvailable()
	if !status.GitAvailable {
		return status
	}

	out, err := runGit("-C", root, "rev-parse", "--is-inside-work-tree")
	if err != nil || strings.TrimSpace(out) != "true" {
		return status
	}
	status.IsRepo = true

	if out, err := runGit("-C", root, "branch", "--show-current"); err == nil {
		br := strings.TrimSpace(out)
		if br != "" {
			status.Branch = &br
		}
	}

	if out, err := runGit("-C", root, "remote", "-v"); err == nil {
		lines := splitLines(out)
		status.Remotes = lines
		status.HasRemote = len(lines) > 0
	}

	if out, err := runGit("-C", root, "status", "--porcelain"); err == nil {
		status.Dirty = strings.TrimSpace(out) != ""
	}

	return status
}

func splitLines(s string) []string {
	s = strings.ReplaceAll(s, "\r\n", "\n")
	parts := strings.Split(s, "\n")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

File: src/main/gitStatus.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import type { GitStatus } from "./types";

const execFileAsync = promisify(execFile);

async function runGit(args: string[], cwd?: string): Promise<{ stdout: string; stderr: string } | null> {
  try {
    const { stdout, stderr } = await execFileAsync("git", args, {
      cwd,
      windowsHide: true,
      maxBuffer: 4 * 1024 * 1024,
    });
    return { stdout, stderr };
  } catch {
    return null;
  }
}

export async function isGitAvailable(): Promise<boolean> {
  const res = await runGit(["--version"]);
  return !!res && res.stdout.toLowerCase().includes("git");
}

export async function getGitStatus(root: string): Promise<GitStatus> {
  const base: GitStatus = {
    gitAvailable: false,
    isRepo: false,
    hasRemote: false,
    remotes: [],
    dirty: false,
    branch: null,
  };

  if (!root) {
    return base;
  }

  const available = await isGitAvailable();
  if (!available) {
    return { ...base, gitAvailable: false };
  }

  const insideRes = await runGit(["-C", root, "rev-parse", "--is-inside-work-tree"]);
  const isRepo = !!insideRes && insideRes.stdout.trim() === "true";

  if (!isRepo) {
    return { ...base, gitAvailable: true, isRepo: false };
  }

  const branchRes = await runGit(["-C", root, "branch", "--show-current"]);
  const branch = branchRes ? branchRes.stdout.trim() || null : null;

  const remoteRes = await runGit(["-C", root, "remote", "-v"]);
  const remotes = remoteRes
    ? remoteRes.stdout
        .split(/\r?\n/)
        .map((l) => l.trim())
        .filter((l) => l.length > 0)
    : [];

  const statusRes = await runGit(["-C", root, "status", "--porcelain"]);
  const dirty = statusRes ? statusRes.stdout.trim().length > 0 : false;

  return {
    gitAvailable: true,
    isRepo: true,
    hasRemote: remotes.length > 0,
    remotes,
    dirty,
    branch,
  };
}
